node/src/lib.rs

```diff
@@ -171,26 +257,99 @@ impl Node {
                 let heartbeat_url = {
                     let nodes = self.model.nodes().await;
                     let node_id = *nodes.iter().next().unwrap().0;
-
-                    let base = &registry.url.join("v1.0/").unwrap();
-                    base.join(&format!("health/nodes/{}", node_id)).unwrap()
+                    let registry = self.current_registry.lock().await.clone().unwrap();
+
+                    registry
+                        .url
+                        .join(&format!("{}/", self.api_version)) // Ensure it ends with a '/'
+                        .unwrap()
+                        .join(&format!("health/nodes/{}", node_id))
+                        .unwrap()
                 };
 
+                let mut first_attempt = true;
                 // Send heartbeat every 5 seconds
                 loop {
+                    debug!("Heart-beating to {}", heartbeat_url);
                     match client.post(heartbeat_url.clone()).send().await {
                         Ok(res) => {
                             if !res.status().is_success() {
-                                error!("Heartbeat error");
+                                if res.status() == StatusCode::NOT_FOUND && first_attempt {
```

Comment:
Try to reregister once if heartbeat produces a 404

Reply:
Ideally the heartbeat is only running after a successful registration ;)

Reply:
its part of the specification, if the registration FAILs we don't heartbeat, if the heartbeat produces a 404 we try to reregister once and try again